### ISSUE TITLE
Can't run austin directly with python scripts

### ISSUE THREAD

**Issue by patrickelectric** [P403n1x87/austin#188]
![image](https://github.com/P403n1x87/austin/assets/1215497/c289cad0-9164-4b53-85c2-03c704d91b09)

**Comment by P403n1x87** [comment]
@patrickelectric is the script executable (i.e. it has a shebang and the +x flag)? If not you need to put `python` or the path to the interpreter executable before the script path.

**Comment by patrickelectric** [comment]
```
root@blueos:~/austin# file /home/pi/services/helper/main.py
/home/pi/services/helper/main.py: Python script, ASCII text executable
root@blueos:~/austin# head /home/pi/services/helper/main.py -n2
#!/usr/bin/env python3
root@blueos:~/austin# ls -lah /home/pi/services/helper/main.py 
Permissions Size User Date Modified Name
.rwxr-xr-x   18k root 21 Aug 23:18  /home/pi/services/helper/main.py
root@blueos:~/austin# austin -i 1ms python /home/pi/services/helper/main.py 
              _   _       
 __ _ _  _ __| |_(_)_ _  
/ _` | || (_-<  _| | ' \ 
\__,_|\_,_/__/\__|_|_||_| 3.5.0 [GCC 10.2.1]

👾  It looks like you are trying to profile a process that is not a Python
process. Make sure that you are targeting the right application. If the Python
process is actually a child of the target process then use the -C option to
discover it automatically.
root@blueos:~/austin# /home/pi/services/helper/main.py 
2023-08-22 11:09:02.501 | INFO     | __main__:<module>:45 - Starting Helper
2023-08-22 11:09:03.426 | DEBUG    | asyncio.selector_events:__init__:54 - Using selector: EpollSelector
```

**Comment by P403n1x87** [comment]
@patrickelectric thanks for confirming. Could you please provide more details about your OS and the CPython version being used? In the meantime, could you also try setting a higher timeout with e.g. `-t 3s` to see if that helps? Thanks 🙏

**Comment by patrickelectric** [comment]
Same result with `-t 3s`. I need to profile the code in the startup process of the script.

```
root@blueos:~# uname -a 
Linux blueos 5.10.92-v7l+ #1514 SMP Mon Jan 17 17:38:03 GMT 2022 armv7l GNU/Linux
root@blueos:~# python --version
Python 3.9.17
```

**Comment by P403n1x87** [comment]
I've just run this little experiment on my machine to try and reproduce
```
gabriele@lima-ubuntu:/tmp$ cat script.py
#!/usr/bin/env python3.9

from time import sleep;sleep(3)

gabriele@lima-ubuntu:/tmp$ austin -i 1ms -o /dev/null ./script.py
              _   _
 __ _ _  _ __| |_(_)_ _
/ _` | || (_-<  _| | ' \
\__,_|\_,_/__/\__|_|_||_| 3.5.0 [GCC 9.4.0]

🐍 Python version: 3.9.16

Statistics
⌛ Sampling duration : 3.03 s
⏱️  Frame sampling (min/avg/max) : 99/455/2812 μs
🐢 Long sampling rate : 9/1016 (0.89 %) samples took longer than the sampling interval to collect
💀 Error rate : 1/1016 (0.10 %) invalid samples
```

Is Python installed from the OS package repository? Does Austin work in general if you run a script using the interpreter in the command (e.g. `austin python script.py`)? If not, could you run
```
nm -gD `which python` | grep _PyRuntime
````
to see if your Python binary has symbols? The output should look like
```
gabriele@lima-ubuntu:/tmp$ nm -gD `which python3.9` | grep _PyRuntime
0000000000997c00 B _PyRuntime
00000000006bab20 T _PyRuntimeState_Fini
00000000006bab80 T _PyRuntimeState_Init
00000000004bfb57 T _PyRuntimeState_ReInitThreads
00000000006c1100 T _PyRuntime_Finalize
00000000006c1120 T _PyRuntime_Initialize
```
Thanks.

**Comment by patrickelectric** [comment]
I just moved on, but I got no output from `nm -gD `which python` | grep _PyRuntime`.

**Comment by P403n1x87** [comment]
> I just moved on, but I got no output from `nm -gD `which python` | grep _PyRuntime`.

sounds like your Python installation has symbols stripped out so that's probably the cause of the issue

**Comment by P403n1x87** [comment]
In this case you might have better luck with [echion](https://github.com/p403n1x87/echion).

**Comment by razorx89** [comment]
I came across the same issue on MacOS Ventura (MBP with M2 CPU). I installed Python 3.11.5 using the official installer from https://www.python.org/downloads/macos/. Invoking `austin` results in the same output as above, even for the simple `time.sleep` example.

```
> austin python3 -c "import time; time.sleep(5)" 
              _   _       
 __ _ _  _ __| |_(_)_ _  
/ _` | || (_-<  _| | ' \ 
\__,_|\_,_/__/\__|_|_||_| 3.5.0 [GCC 4.2.1]

👾 It looks like you are trying to profile a process that is not a Python
process. Make sure that you are targeting the right application. If the Python
process is actually a child of the target process then use the -C option to
discover it automatically.
```

`nm` does not yield any information:
```
nm -gD `which python3.11`                  
/Library/Developer/CommandLineTools/usr/bin/nm: error: /usr/local/bin/python3.11: File format has no dynamic symbol table
```

**Comment by P403n1x87** [comment]
@razorx89 I think your issue is different. If you have installed Python with the official installer, then the issue is that the binaries are signed. The "solution" in this case is to remove the signature (see https://github.com/P403n1x87/austin#on-macos-1 for details). To confirm that the binaries have symbols you would have to `nm -gD` the library rather than the executable, e.g. something like
```
nm -gD /Library/Frameworks/Python.framework/Versions/3.11/Resources/Python.app/Contents/MacOS/Python
```

**Comment by razorx89** [comment]
Thanks for the quick response. Unfortunately, removing the code signature as described prevents the Python executable from being started. The process is getting killed directly.

**Comment by sihil** [comment]
I've just come across this thread trying to get Austin working with a pyenv installed python 3.8. I then discovered that pyenv builds don't have any symbols in them until 3.12. See https://github.com/pyenv/pyenv/issues/2466 for more details.

**Comment by P403n1x87** [comment]
@sihil I use pyenv for local testing on MacOS
```
❯ pyenv versions       
* system (set by /Users/gabriele/.pyenv/version)
* 2.7.18 (set by /Users/gabriele/.pyenv/version)
* 3.6.15 (set by /Users/gabriele/.pyenv/version)
* 3.7.16 (set by /Users/gabriele/.pyenv/version)
* 3.8.16 (set by /Users/gabriele/.pyenv/version)
* 3.9.16 (set by /Users/gabriele/.pyenv/version)
* 3.10.9 (set by /Users/gabriele/.pyenv/version)
* 3.11.1 (set by /Users/gabriele/.pyenv/version)
* 3.12.0rc1 (set by /Users/gabriele/.pyenv/version)
```
and symbol resolution in Austin seems to work as expected. In fact:
```
❯ nm -gU /Users/gabriele/.pyenv/versions/3.8.16/lib/libpython3.8.dylib | grep __PyRuntime
00000000002ac3d8 S __PyRuntime
000000000014a8d0 T __PyRuntimeState_Fini
000000000014a7b0 T __PyRuntimeState_Init
000000000014a930 T __PyRuntimeState_ReInitThreads
0000000000147a00 T __PyRuntime_Finalize
00000000001479c0 T __PyRuntime_Initialize
```
This is from the logs of Austin 3.5.0 running against Python 3.8 with `sudo src/austin -t 1s python3.8 -c "from time import sleep;sleep(4)"`
```
2023-09-08 19:33:09.431 Db austin[95837:ec8813] Symbol __PyRuntime found @ 0x1039383d8
2023-09-08 19:33:09.431 Db austin[95837:ec8813] Library map: /Users/gabriele/.pyenv/versions/3.8.16/lib/libpython3.8.dylib (with symbols)
2023-09-08 19:33:09.431 Db austin[95837:ec8813] BSS bounds  [0x103918850 - 0x1039382e0]
2023-09-08 19:33:09.431 Db austin[95837:ec8813] HEAP bounds [0x1032d8000 - 0x92b15c000]
2023-09-08 19:33:09.431 Db austin[95837:ec8813] Process is ready
2023-09-08 19:33:09.431 Db austin[95837:ec8813] _py_proc__init timer loop terminated
2023-09-08 19:33:09.431 Db austin[95837:ec8813] Python binary:  /Users/gabriele/.pyenv/versions/3.8.16/bin/python3.8
2023-09-08 19:33:09.431 Db austin[95837:ec8813] Python library: /Users/gabriele/.pyenv/versions/3.8.16/lib/libpython3.8.dylib
2023-09-08 19:33:09.431 Db austin[95837:ec8813] Maximal VM address space: 0x1032d8000-0x7ffffffc5000
2023-09-08 19:33:09.431 Db austin[95837:ec8813] Python version (from file name): 3.8.0
2023-09-08 19:33:09.432 Db austin[95837:ec8813] GC runtime state @ 0x103938530
2023-09-08 19:33:09.432 Db austin[95837:ec8813] Found possible interpreter state @ 0x7fe08700a200 (offset 0x7fdf83d29200).
2023-09-08 19:33:09.432 Db austin[95837:ec8813] \-'\-( Interpreter head de-referenced from symbols \-'\-(
2023-09-08 19:33:09.432 Db austin[95837:ec8813] Interpreter State de-referenced @ raddr: 0x7fe08700a200 after 1 attempts
```
Could you open a new issue with all the details of the problem that you are facing, please? 🙏

**Comment by razorx89** [comment]
> @razorx89 I think your issue is different. If you have installed Python with the official installer, then the issue is that the binaries are signed. The "solution" in this case is to remove the signature (see https://github.com/P403n1x87/austin#on-macos-1 for details). To confirm that the binaries have symbols you would have to `nm -gD` the library rather than the executable, e.g. something like
> 
> ```
> nm -gD /Library/Frameworks/Python.framework/Versions/3.11/Resources/Python.app/Contents/MacOS/Python
> ```

After seeing your other response, yes the library has symbols:
```
> nm -gU /Library/Frameworks/Python.framework/Versions/3.11/lib/libpython3.11.dylib | grep __PyRuntime
00000000004e6258 S __PyRuntime
0000000000217218 T __PyRuntimeState_Fini
0000000000217068 T __PyRuntimeState_Init
000000000020d0d4 T __PyRuntime_Finalize
000000000020d0a4 T __PyRuntime_Initialize
```

But the part with removing the code signature still does not work. Can you please clarify what the appropriate steps are to run Python (with/without austin) after removing the code signature? Sorry, I don't have any experience in messing up the security layer of MacOS :)

**Comment by P403n1x87** [comment]
@razorx89 In my experience, removing the signature from the binaries does not affect their ability to execute. This seems to work fine in CI

https://github.com/P403n1x87/austin/blob/b7eb5e2ba6c500b59a4cf98a3193be47c01fe5f3/.github/workflows/tests.yml#L156C24-L157

and I have also done this locally in the past, albeit on an Intel Mac and perhaps not the latest available OS version. One thing to check is the Security Settings, whether the python binaries shows up as blocked and wanting permission to run (see e.g. [this SE post](https://apple.stackexchange.com/questions/436674/how-to-unblock-binary-from-use-because-mac-says-it-is-not-from-identified-develo))

**Comment by tjdcs** [comment]
I believe I'm having the same issue. 

Python installed with homebrew
```
$ brew install python@3.11
$ python -V -V
Python 3.11.6 (main, Oct  2 2023, 13:45:54) [Clang 15.0.0 (clang-1500.0.40.1)]
```

Running austin 3.6.0 as sudo with suduers edited gives me the same message about not being attached to a python processes. 

```
(.venv) ➜  colour-workbench git:(main) ✗ nm -gD `which python`    
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/nm: error: /Users/tucker/Dev/colour-workbench/.venv/bin/python: File format has no dynamic symbol table
```

No symbol table...

I think the issue is specifically with MacOS sonoma (14.0) which probably has some security updates. Opps. I wish I hadn't upgraded.

**Comment by P403n1x87** [comment]
@tjdcs thanks for your report. May I get some more details from you about your issue, please?
- The homebrew version of Python should be installed in an area that I believe is not covered (at least pre-sonoma) by special system protections. The installation might still come with a "libpython" which would be the binary containing the symbols. Would you be able to check whether that's the case? Note that the argument to `nm` should be `-gU` rather than `-gD` (or simply `-g`).
- How was Austin installed? This shouldn't matter much but it's just for getting the full picture.
- What command are you running? Are you trying to attach to a running Python process or launching one with Austin?

I haven't tested Austin on sonoma, and I believe the current CI runners don't quite capture the state of the latest OSX release. A few things to try, if possible, are:
- Install Python with pyenv; this is what I use for local tests. CI uses the official Python installers, which requires the signature to be removed from the binaries.
- Run Austin on a Linux machine, e.g. [lima](https://github.com/lima-vm/lima).
- Try [Echion](https://github.com/P403n1x87/echion). Attaching to a running process might have the same issues as Austin permissions-wise, but launching a command does not require sudo because of the in-process nature of the tool.

**Comment by tjdcs** [comment]
```
$ nm -gU `which python`   
0000000100000000 T __mh_execute_header
```

I tried removing the signature on the underlying binary but then when I run a python script it looks like the OS immediately kills the process. 

😬 bad mac update for sure. Big mistake on my part 😂

**Comment by tjdcs** [comment]
echion does work without issue. It just doesn't accept the --pipe argument and the austin vs-code plugin actions aren't seemless to use. 

Easy enough to work around though. 

I'll work with echion for now. Thanks for your work on these great tools!!! A definite friend of https://github.com/colour-science/colour

**Comment by P403n1x87** [comment]
> I'll work with echion for now. Thanks for your work on these great tools!!! A definite friend of https://github.com/colour-science/colour

Glad Echion offered a workaround for now! It's currently pretty much an experiment and still playing catch-up with Austin features.
